torch/csrc/jit/codegen/cuda/lower_alias_memory.cpp

```diff
@@ -851,50 +925,61 @@ void BufferReuseDebugPrinter::printAllocInfo(const kir::Allocate* alloc) {
   if (alloc_info->can_use_inner_alias) {
     os_ << "inner live interval: ";
     os_ << alloc_info->inner_live_interval->toString() << " , ";
+  } else {
+    os_ << "cannot use inner alias, ";
   }
   os_ << "size expr : " << alloc_info->size_expr << " , "
       << "outer live interval: " << alloc_info->outer_live_interval->toString();
   indent() << message_end;
 }
 
 //! Reuse Allocation nodes via pointer aliasing
-class AllocateReuseModifier {
+class ReusableAllocationFinder : private kir::IrVisitor {
```

Comment:
See an issue here: https://github.com/csarofeen/pytorch/issues/1785, but I am still concerned. I don't think broadcasting resolution is the only case we should worry about. For example, I can have code:
```C++
float T1[2];
compute T1[2]
for i in range(2):
  something = T1[i];
  // no more read of T1 below this point.
  float T2[2];  // this will reuse T1
  read T2 from fusion input
  something_else = reduction(T2)
  fusion_output[index] = something + something_else
```

Reply:
I think this case would be covered by `isSerialBroadcastResolution`, but not 100% sure. Can you create a repro and file an issue? I want to focus just refactoring in this PR. There's one bug I'm fixing in #2174.

Reply:
Yeah, you didn't change the algorithm in this PR, so this is not a blocking comment. I am just generally worried about the algorithm and will try to think if I can break it.

Reply:
OK, I think the above case I mentioned is covered in
```C++
  //! Careful heavy check on inner sharing candidates,
  //!  current enforced conditions are:
  //!
  //! 1. The two buffers have producer-consumer relationship
```